dan_layer/core/src/workers/pacemaker_worker.rs

```diff
@@ -104,11 +110,15 @@ where T: Clone + Debug + PartialEq + Eq + Hash + Send + Sync + 'static
                         "Received start wait signal for value: {:?}", wait_over
                     );
                     let (tx, rx_stop_signal) = oneshot::channel();
-                    if self.pending_timeouts.insert(wait_over.clone(), tx).is_none() {
+                    let pending_timeouts = self.pending_timeouts.clone();
+                    if pending_timeouts.lock().unwrap().insert(wait_over.clone(), tx).is_none() {
                         self.waiting_futures.push(Box::pin(async move {
                             tokio::select! {
                                 _ = tokio::time::sleep(duration_timeout) => {
                                     info!("The wait signal for value = {:?} has timeout", wait_over);
+                                    // remove the wait_over from the pending_timeouts
+                                    // so that this data structure doesn't grow every time
+                                    pending_timeouts.lock().unwrap().remove(&wait_over);
```

Comment:
Sync locks in async code are not good.

You shouldnt need a lock, the waiting_future will resolve one way or another so you should be able to  clean up pending timeouts then

Reply:
Oh I see, that is actually a clever approach. Let me refactor it, to put these considerations in place.